beacon_node/eth1/src/service.rs

```diff
@@ -37,8 +41,214 @@ const WARNING_MSG: &str = "BLOCK PROPOSALS WILL FAIL WITHOUT VALID ETH1 CONNECTI
+        if let Some(result) = *endpoint.1.read().await {
+            return result;
+        }
+        let mut value = endpoint.1.write().await;
+        if let Some(result) = *value {
+            return result;
+        }
+        crate::metrics::inc_counter_vec(&crate::metrics::ENDPOINT_REQUESTS, &[&endpoint.0]);
+        let state = endpoint_state(
+            &endpoint.0,
+            &self.config_network_id,
+            &self.config_chain_id,
+            &self.log,
+        )
+        .await;
+        *value = Some(state);
+        if state.is_err() {
+            crate::metrics::inc_counter_vec(&crate::metrics::ENDPOINT_ERRORS, &[&endpoint.0]);
+        }
+        state
+    }
+
+    pub async fn first_success<'a, F, O, R>(
+        &'a self,
+        func: F,
+    ) -> Result<O, FallbackError<SingleEndpointError>>
+    where
+        F: Fn(&'a str) -> R,
+        R: Future<Output = Result<O, SingleEndpointError>>,
+    {
+        let func = &func;
+        self.fallback
+            .first_success(|endpoint| async move {
+                match self.state(endpoint).await {
+                    Ok(()) => {
+                        let endpoint_str = &endpoint.0;
+                        crate::metrics::inc_counter_vec(
+                            &crate::metrics::ENDPOINT_REQUESTS,
+                            &[endpoint_str],
+                        );
+                        match func(&endpoint.0).await {
+                            Ok(t) => Ok(t),
+                            Err(t) => {
+                                crate::metrics::inc_counter_vec(
+                                    &crate::metrics::ENDPOINT_ERRORS,
+                                    &[endpoint_str],
+                                );
+                                if let SingleEndpointError::EndpointError(e) = &t {
+                                    *endpoint.1.write().await = Some(Err(*e));
+                                }
+                                Err(t)
+                            }
+                        }
+                    }
+                    Err(e) => Err(SingleEndpointError::EndpointError(e)),
+                }
+            })
+            .await
+    }
+}
+
+/// Returns `Ok` if the endpoint is usable, i.e. is reachable and has a correct network id and
+/// chain id. Otherwise it returns `Err`.
+async fn endpoint_state(
+    endpoint: &str,
+    config_network_id: &Eth1Id,
+    config_chain_id: &Eth1Id,
+    log: &Logger,
+) -> EndpointState {
+    let error_connecting = |_| {
+        warn!(
+            log,
+            "Error connecting to eth1 node. Trying fallback ...";
+            "endpoint" => endpoint,
+        );
+        EndpointError::NotReachable
+    };
+    let network_id = get_network_id(endpoint, Duration::from_millis(STANDARD_TIMEOUT_MILLIS))
+        .await
+        .map_err(error_connecting)?;
+    if &network_id != config_network_id {
+        warn!(
```

Comment:
This is in conflict with #1981 unfortunately. I *think* what we need to do is return some error if we get `chain_id ==0` so that we try again.

Reply:
I am returning now an `EndpintError::FarBehind` error + logging a warning.